proxy.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Log forwarders like fluentd or fluentbit post log lines in bulk to this
+// elasticsearch endpoint
+const BulkLogsEndpoint = "/_bulk"
+
+func unauthorized(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusUnauthorized)
+	w.Write([]byte("{\"message\":\"unauthorized\"}"))
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+type Proxy struct {
+	config        *ProxyConfig
+	httpClient    *http.Client
+	remoteURL     *url.URL
+	signer        Signer
+	authenticator Authenticator
+}
+
+// isValidTargetEndpoint checks if the request is has a valid target depending
+// on the configuration of the proxy
+func (p *Proxy) isValidTargetEndpoint(r *http.Request) bool {
+	if p.config.LogsOnly {
+		if r.URL.Path != BulkLogsEndpoint {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !p.isValidTargetEndpoint(r) {
+		unauthorized(w)
+		return
+	}
+
+	if p.config.InputValidation {
+		ok := p.authenticator.AuthenticateRequest(r)
+		if !ok {
+			msg := "not authorized to proxy request"
+			http.Error(w, msg, http.StatusUnauthorized)
+			return
+		}
+	}
+
+	if p.config.OutputSigning {
+		err := p.signer.SignRequest(r)
+		if err != nil {
+			msg := "could not sign request"
+			http.Error(w, msg, http.StatusInternalServerError)
+			log.Error(err)
+			return
+		}
+	}
+
+	// RequestURI should not be set, otherwise http.Client witll throw an
+	// error
+	r.RequestURI = ""
+	r.URL = p.remoteURL
+
+	p.ProxyRequest(w, r)
+}
+
+func (p *Proxy) ProxyRequest(w http.ResponseWriter, r *http.Request) {
+	resp, err := p.httpClient.Do(r)
+	if err != nil {
+		http.Error(w, "Server Error", http.StatusInternalServerError)
+		log.Error("ServeHTTP:", err)
```

Comment:
Hm, we might be losing some information contained within custom error structs. Some custom error types add additional parameter fields with information helpful during debugging. Think it would be safer to use `log.WithField("error", err).Error("Proxy'ing request failed")`. By simply passing `err` into `Error()`, it's `.toString()` method is being called on it, and being outputted to the logs as a simple string.

Reply:
moved to `log.WithField`